### PR TITLE
Fix nullable entity comparison with null

### LINKED ISSUE #2611: One-to-zero-or-one relation not returning data when checking for null
# Issue/Scenario

We are having troubles with some of our queries that contain null checks and don't return any data when the null check is in place. Our setup is as follows:

- database first (legacy database)
- one-to-zero-or-one relationship between two tables
- some queries contain null checks on the zero-or-one side and do not return data because of the null check

We are not sure if this is 

- expected behavior/by design (zero-to-one-side mapped with Id generated by foreign)
- wrong mapping on our side
- a bug

## Steps to reproduce

- Query data:
  - `.Query<Sheep>().Where(s => s.Name == "Dolly")` 
  - **Result: *** Sheep "Dolly" returned
- Add a null check to the query (zero-or-one-side):
  -  `.Query<Sheep>().Where(s => s.Name == "Dolly" && s.SlaughterInfo == null)` 
  - **Result: *** no data returned due to generated SQL statement containing ` (slaughterInfo.SheepId is null)` 

I have created a [sample project](https://github.com/marcelstuder/nhibernate-one-to-zero-or-one-issue) that demonstrates the issue.

Even if this is the expected behavior: from a developer's perspective who writes the Linq query it is not obvious that this null check is causing the query to fail without looking at the underlying mapping (or writing a test :-)).

## Notes

We are using the following Versions:

- NHibernate 5.3.5
- FluentNhibernate 3.1.0

## Workarounds

- check all queries in code and remove null checks if possible

### PR DESCRIPTION
**Pull request by bahusoid**
Fixes #2611

### CODE REVIEW

**Comment by fredericDelaporte** [comment]
So this would somehow fix #2611 by reverting to pre #2081 behavior for a special and narrow case?

**Comment by bahusoid** [comment]
>  this would somehow fix #2611

No as one-to-one case remains completely broken but with different results as Id of the owner is used in this case. By example:
```C#
s.Query<Parent>().Where(p => p.OneToOne == null);
```

5.2 (and 5.3.10 with this PR applied):
```sql
select * from Parent p where p.Id is null --Returns all parent records
```

5.3.9 :
```sql
select * from Parent p, OneToOne  o where p.Id = o.Id and o.Id is null -- Returns zero records
```

**Comment by hazzik** [comment]
Do I understand correctly that the correct SQL would be like this?

```sql
select * from Parent p left join OneToOne  o on p.Id = o.Id where o.Id is null -- Returns zero records
```

**Comment by bahusoid** [comment]
Yes. Left join is needed instead of implicit

**Comment by hazzik** [comment]
image

I made it working. Need to resolve TODO. Feel free to drop the commit if you don't like it.

File: src/NHibernate/Hql/Ast/ANTLR/Tree/DotNode.cs [hunk 1]
```diff
@@ -417,8 +415,15 @@ private void DereferenceEntity(EntityType entityType, bool implicitJoin, string
 				               || comparisonWithNullableEntity;
 			}
 
-			if ( joinIsNeeded ) 
+			if ( joinIsNeeded )
 			{
+				//TODO: Find correct place for this if.
+				if (Walker.IsNullComparison)
```
Comment (hazzik) [inline src/NHibernate/Hql/Ast/ANTLR/Tree/DotNode.cs:421]: Need to find correct place for this check. It probably needs to be done somewhere earlier. I don't have time right now, could look later, but not promising it.

**Comment by bahusoid** [comment]
> I made it working.

I made similar changes in https://github.com/nhibernate/nhibernate-core/pull/2612. Only then I tried to make it more general fix not bound to is null comparison. Yeah I also realized that merging both PRs might be a good idea..

**Review by hazzik — APPROVED** [review]

**Review by hazzik — APPROVED** [review]

**Pull request merged**